src/blocks/packages.rs

```diff
@@ -0,0 +1,351 @@
+    let format_singular = config
+        .format_singular
+        .with_default(" $icon $total.eng(w:1) ")?;
+    let format_up_to_date = config
+        .format_up_to_date
+        .with_default(" $icon $total.eng(w:1) ")?;
+
+    // If user provide package manager in any of the formats then consider that also
+    macro_rules! any_format_contains {
+        ($name:expr) => {
+            format.contains_key($name)
+                || format_singular.contains_key($name)
+                || format_up_to_date.contains_key($name)
+        };
+    }
+
+    let apt = any_format_contains!("apt");
+    let aur = any_format_contains!("aur");
+    let pacman = any_format_contains!("pacman");
+
+    if !config.package_manager.contains(&PackageManager::Apt) && apt {
+        config.package_manager.push(PackageManager::Apt);
+    }
+    if !config.package_manager.contains(&PackageManager::Pacman) && pacman {
+        config.package_manager.push(PackageManager::Pacman);
+    }
+    if !config.package_manager.contains(&PackageManager::Aur) && aur {
+        config.package_manager.push(PackageManager::Aur);
+    }
+
+    let warning_updates_regex = config
+        .warning_updates_regex
+        .as_deref()
+        .map(Regex::new)
+        .transpose()
+        .error("invalid warning updates regex")?;
+    let critical_updates_regex = config
+        .critical_updates_regex
+        .as_deref()
+        .map(Regex::new)
+        .transpose()
+        .error("invalid critical updates regex")?;
+    let ignore_updates_regex = config
+        .ignore_updates_regex
+        .as_deref()
+        .map(Regex::new)
+        .transpose()
+        .error("invalid ignore updates regex")?;
+
+    let mut package_manager_vec = Vec::new();
+
+    for &package_manager in config.package_manager.iter() {
+        let backend = match package_manager {
+            PackageManager::Apt => Box::new(Apt::new().await?) as Box<dyn Backend>,
+            PackageManager::Pacman => Box::new(Pacman::new().await?) as Box<dyn Backend>,
+            PackageManager::Aur => Box::new(Aur::new()) as Box<dyn Backend>,
+        };
+
+        package_manager_vec.push(backend);
+    }
+
+    loop {
+        let (mut apt_count, mut pacman_count, mut aur_count) = (0, 0, 0);
+        let mut critical = false;
+        let mut warning = false;
+
+        // Iterate over the all package manager listed in Config
+        for package_manager in &package_manager_vec {
+            let updates = package_manager.get_updates_list().await?;
+
+            match package_manager.package_manager() {
+                PackageManager::Apt => {
+                    let apt = Apt {
+                        // Config file not needed in counting updates
+                        config_file: String::new(),
+                        ignore_phased_updates: config.ignore_phased_updates,
+                        ignore_updates_regex: ignore_updates_regex.clone(),
+                    };
```

Comment:
- Why not pass these parameters once when creating the backend?
- `ignore_updates_regex` is not apt-specific and should be applied to all backends.

Reply:
> Why not pass these parameters once when creating the backend?

Done 

> `ignore_updates_regex` is not apt-specific and should be applied to all backends.

Pacman and Aur doesn't need them anywhere while checking updates or counting them
That's why the Pacman & Aur struct don't have them as variant

Reply:
Neither does apt. This is a generic user setting to filter the updates they don't care about. See #1958.

Reply:
Does` _ignore_updates_regex` allowed because that parameter is never read

Reply:
Not sure what you are referring to. Don't store it in every driver, just apply it when iterating over package managers if available.

Reply:
> Don't store it in every driver, just apply it when iterating over package managers if available.

Reply:
```rust
Box::new(Pacman::new(ignore_updates_regex.clone()).await?) as Box<dyn Backend>
```
You mean here just send `ignore_updates_regex` through `Pacman::new()` but don't store it as a variant or do anything else

Reply:
No, backends don't have to know about `ignore_updates_regex` at all. When you call `get_updates_list()`, check if `ignore_updates_regex` is set, and it if is, filter the list.

Reply:
Do you need something like this
```rust
if let Some(_) = ignore_updates_regex {
    let updates_count = package_manager
        .get_update_count(&updates, ignore_updates_regex)
        .await?;
}
```

And then `get_count_update` will filter accordingly

Reply:
No, I was thinking about something like

```rust
let updates = package_manager.get_updates().await?;
if let Some(re) = ignore_updates_regex {
	updates.retain(|u| !re.is_match(u));
}
```

Here `.get_updates()` returns a `Vec<String>` though, so each driver should convert a string to a list of lines, applying filtering if necessary.

Reply:
Oh okk

Reply:
Done